x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptFieldMapper.java

```diff
@@ -56,81 +74,122 @@ protected String contentType() {
+                    if (optionsObject instanceof Map == false) {
+                        throw new IllegalArgumentException("unable to parse options for script field [" + name + "]");
+                    }
+                    options = (Map<String, String>) optionsObject;
+                } else {
+                    options = Collections.emptyMap();
+                }
+                if (scriptMap.size() > 0) {
+                    throw new IllegalArgumentException(
+                        "unsupported parameters specified for script field [" + name + "]: " + scriptMap.keySet()
+                    );
+                }
+                return new Script(ScriptType.INLINE, PainlessScriptEngine.NAME, sourceObject.toString(), options, params);
+            } else if (scriptObject instanceof String) {
+                return new Script((String) scriptObject);
+            } else {
+                throw new IllegalArgumentException("unable to parse script for script field [" + name + "]");
 
-        public void setScriptService(ScriptService scriptService) {
-            this.scriptService.set(scriptService);
+            }
         }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
 
         @Override
         public ScriptFieldMapper.Builder parse(String name, Map<String, Object> node, ParserContext parserContext)
             throws MapperParsingException {
-            Builder builder = new Builder(name, scriptService.get());
-            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
-                Map.Entry<String, Object> entry = iterator.next();
-                String propName = entry.getKey();
-                Object propNode = entry.getValue();
-                if (propName.equals("runtime_type")) {
-                    if (propNode == null) {
-                        throw new MapperParsingException("Property [runtime_type] cannot be null.");
-                    }
-                    builder.runtimeType(XContentMapValues.nodeStringValue(propNode, name + ".runtime_type"));
-                    iterator.remove();
-                } else if (propName.equals("script")) {
-                    if (propNode == null) {
-                        throw new MapperParsingException("Property [script] cannot be null.");
-                    }
-                    // TODO this should become an object and support the usual script syntax, including lang and params
-                    builder.script(new Script(XContentMapValues.nodeStringValue(propNode, name + ".script")));
-                    iterator.remove();
-                }
-            }
-            // TODO these get passed in sometimes and we don't need them
-            node.remove("doc_values");
-            node.remove("index");
+            ScriptFieldMapper.Builder builder = new ScriptFieldMapper.Builder(name, parserContext.queryShardContextSupplier());
```

Comment:
I think `queryShardContextSupplier` exists for the percolator and it is a bit sneaky. It'll do but I'd feel more comfortable if we had the `ScriptService` directly.